Cleanup and speed improvement of MapHeapHogTest

Donnerbart opened the pull request:
Cleanup and speed improvement of MapHeapHogTest (by selecting smarter keys for each local thread).

Danny-Hazelcast commented:
Originally I was using this test,  there the max heap used factor, was a precondition to starting the test.
and so this test was not really fitting in with other loop/time based tests.   originally I was waiting until the each member in cluster used up its heap,  (which is tacking some unknown amount of time)  and then doing some thing.   Now we are shoehorning the test into a new "test format"  and really changing the meaning of this test.

Donnerbart commented:
I don't think that I've changed the behavior of the test in a significant way. It was looping in a `while (!testContext.isStopped())` loop before, as it is now. It just moved the logging into the inner `for` loop, so it actually has a change to log something during runtime.

I also changed the key selection if run with multiple threads (before: all the same key, now: all different keys). If the load due to several threads and the slow speed was intended, I'm sorry for changing this. This could be configurable easily (it's just the `increment` value of the `nextKeyOwnedBy` method, the different `localStartKey` doesn't matter at all for a long running test).

Donnerbart commented:
The default behavior (regarding the key selection) should be restored. Everything else should not change the behavior, also the default values are the same.

New property `useSameKeysOnAllWorker` to make each worker use a separate key (which speeds up the fill rate according to the number of worker threads).

New property `stopWorkerOnTestContextStop` to stop the inner loop early, if the test context has been stopped.

New property `stopWorkerAfterFirstTimeStep` to stop the workers after the first fill of the heap.

Danny-Hazelcast commented on tests/src/main/java/com/hazelcast/stabilizer/tests/helpers/HazelcastTestUtils.java:
```diff
@@ -167,16 +167,32 @@ public static HazelcastInstanceImpl getHazelcastInstanceImpl(HazelcastInstance h
+     * @return next key owned by given Hazelcast instance
+     * @throws RuntimeException if no key could be found after 100 retries
+     */
+    public static long nextKeyOwnedBy(HazelcastInstance instance, long lastKey, int increment) {
```
we we really need to complicate this method with an increment,  which has not real meaning.   we just want to return the next long that is owned by the instance,  (inclusive of LastKey value)

Donnerbart replied:
Hmm, if we don't use an `increment > 1` we cannot reasonably use more than one worker. Otherwise all will share the same key, since the method has no synchronization between the threads (which is good for performance).

If you just want to use multiple threads to increase load on the cluster, we should make that more obvious. Otherwise the usage of this method looks like that all workers are filling the cluster in parallel with different keys. But in fact they don't.

Danny-Hazelcast commented on tests/src/main/java/com/hazelcast/stabilizer/tests/helpers/HazelcastTestUtils.java:
```diff
@@ -167,16 +167,32 @@ public static HazelcastInstanceImpl getHazelcastInstanceImpl(HazelcastInstance h
+        Member localMember = instance.getCluster().getLocalMember();
+        PartitionService partitionService = instance.getPartitionService();
+        long key = lastKey;
+        for (int retries = 0; retries < 100; retries++) {
```
I personally like the loop for ever trick for this, yes it loops for ever but,  in reality it should not.
also 100 is hard coded, and a very small number of retries.  if we have a 256 node cluster,  I might not own any keys in the range 0 to 100

Danny-Hazelcast commented on tests/src/main/java/com/hazelcast/stabilizer/tests/helpers/HazelcastTestUtils.java:
```diff
@@ -167,16 +167,32 @@ public static HazelcastInstanceImpl getHazelcastInstanceImpl(HazelcastInstance h
-            key++;
+            key += increment;
         }
+        throw new RuntimeException("Could not find next key for " + lastKey + " with increment " + increment);
```
could not find key owned by instance, in the range lastKey to MAX_RETRIES inclusive

Danny-Hazelcast commented on tests/src/main/java/com/hazelcast/stabilizer/tests/map/MapHeapHogTest.java:
```diff
@@ -1,116 +1,149 @@
+    /**
+     * Configures the worker threads to use the same keys, which slows down the fill rate while keeping the load high.
+     */
+    public boolean useSameKeysOnAllWorker = true;
```
I don't understand how this works 

the logic of the old test was to specify a max number of keys we wanted, in cluster.    agnostic to cluster size.

each member of the cluster would then loop through the MAX number of keys  putting only keys that was owned by the  member.   

this way we can control the number of keys in the map precisely, and fill up the map in the quickest possible time,  and all puts are local (only needing network for  backup's),  

I know the loop in this test is controlled by the heap space remaining,  but the logic is the same.  so how do you think, we are making this process faster, as per the title of the pr ?

Donnerbart replied:
`threadCount` was set to 3, so we had 3 local worker threads per member, but all were using the same local keys (start with 0, increase by 1):
```
for (i < maxKeys) {
  key = 0;
  map.put(key, key);
  findNextLocalKey(key);
  key += 1;}
```

So by increasing the `threadCount` we just spawned multiple workers which did exactly the same thing. Now we can configure (or change this as default) that each worker should use a different local key (start with 0 -> threadCount, increase by threadCount):
```
for (i < maxKeys) {
  key = atomicInt.getAndIncrement();
  map.put(key, key);
  findNextLocalKey(key, threadCount);
  key += threadCount }
```

This will speed up the inserts. If you did not intend to just create more load on the cluster with multiple workers, we can make this the default behavior and drop that property for that.

Donnerbart replied:
Let's make this easy: You want to fill the cluster map with data (by putting local keys into the map).

Should the test do this as fast as possible or by creating the most possible load?

Speed: all local worker threads should use different keys (fast + parallel puts)
Load: all local worker threads should use the same keys (slow + parallel puts)

Danny-Hazelcast commented on tests/src/main/java/com/hazelcast/stabilizer/tests/map/MapHeapHogTest.java:
```diff
@@ -1,116 +1,149 @@
+
+            long key = localStartKey;
+            for (int workerIteration = 0; workerIteration < maxIterations; workerIteration++) {
+                key = nextKeyOwnedBy(targetInstance, key, (useSameKeysOnAllWorker ? 1 : threadCount));
```
complicated ternary inside function call

Danny-Hazelcast commented on tests/src/main/java/com/hazelcast/stabilizer/tests/map/MapHeapHogTest.java:
```diff
@@ -1,116 +1,149 @@
+            long key = localStartKey;
+            for (int workerIteration = 0; workerIteration < maxIterations; workerIteration++) {
+                key = nextKeyOwnedBy(targetInstance, key, (useSameKeysOnAllWorker ? 1 : threadCount));
+                map.put(key, key, ttlHours, TimeUnit.HOURS);
```
then a simple put if absent hear,  to make sure multi threads running on one member will only put the value once

Donnerbart replied:
This makes no sense. You just create a huge race between multiple worker threads, which will all try to insert the same key. The `putIfAbsent` will not change anything, but increase the load on the cluster, due to the additional checks.

Danny-Hazelcast commented on tests/src/main/java/com/hazelcast/stabilizer/tests/map/MapHeapHogTest.java:
```diff
@@ -1,116 +1,149 @@
+            long maxIterations = (long) ((double) maxLocalEntries / (double) threadCount);
+
+            long key = localStartKey;
+            for (int workerIteration = 0; workerIteration < maxIterations; workerIteration++) {
```
whats all the workerItterations,  cant we just go round all the keys and put them

Danny-Hazelcast commented:
also as an note,  originally I wanted to do the, the filling up of the map in the warmup or setup phase of the test,  however there is a default time out,  which caused problems,  So I just put it  into the run method of this test.  The purpose of this test, was to create a cluster where each member was using a high proportion of heap space.  then terminate one of the members.  this leads to a cluster melt down and all members crash.  I was not interested in, printing out messages a iteration, or measurint any kind of performance of any thing,  its purely for, this melt down senario.  so the best thing for this test would be to remove the timeout in the warmup / setup phase,  and leave the run method empty

Donnerbart commented:
Okay, so you want to fill the heap as fast as possible in the warmup phase :)

Donnerbart commented:
Okay, moved everything to the warmup phase, empty run is in place. Each member fills the cluster map with 10 threads per default. Since every thread uses its own key range the map will be filled quite fast.

This timeout thing is very well hidden in `CommandFuture.get()` call, if i figured this out correctly. I'll see if I can find a good solution for this, e.g. with a parameter in the `@Warmup` annotation.

pveentjer commented on tests/src/main/java/com/hazelcast/stabilizer/tests/map/MapHeapHogWarmupTest.java:
```diff
@@ -0,0 +1,136 @@
+        maxEntriesPerThread = (long) ((double) maxLocalEntries / (double) threadCount);
+    }
+
+    @Warmup()
```
Do we need those () ?

pveentjer commented on tests/src/main/java/com/hazelcast/stabilizer/tests/map/MapHeapHogWarmupTest.java:
```diff
@@ -0,0 +1,136 @@
+ */
+public class MapHeapHogWarmupTest {
+
+    private static final int MIN_CLUSTER_SIZE = 3;
```
Why do we want hard coded cluster sizes in the test?

Donnerbart replied:
I can understand the intention. All operations are executed on local keys, so we don't want a changing cluster during the execution. So there is this wait check in the warmup.

I don't know if this is necessary though. Isn't the cluster already stable in the warmup phase? So maybe we can skip this whole part of waiting for that.

pveentjer commented on tests/src/main/java/com/hazelcast/stabilizer/tests/map/MapHeapHogWarmupTest.java:
```diff
@@ -0,0 +1,136 @@
+        }
+
+        // wait for the cluster to build up
+        while (targetInstance.getCluster().getMembers().size() != MIN_CLUSTER_SIZE) {
```
I think we should have some kind of timeout. You don't want to wait till the end of time.

pveentjer commented on tests/src/main/java/com/hazelcast/stabilizer/tests/map/MapHeapHogWarmupTest.java:
```diff
@@ -0,0 +1,136 @@
+    }
+
+    @Run
+    public void run() {
```
Why is nothing done in the run phase?? I'm sure there is a reason, but I don't know if I'm going to like this one. This is not how stabilizer is intended to be used.

Donnerbart replied:
Again, I can understand the intention. See this quote from Danny

> The purpose of this test, was to create a cluster where each member was using a high proportion of heap space. then terminate one of the members. this leads to a cluster melt down and all members crash. I was not interested in, printing out messages a iteration, or measurint any kind of performance of any thing, its purely for, this melt down senario. so the best thing for this test would be to remove the timeout in the warmup / setup phase, and leave the run method empty

Maybe we should move this to distinct package for `non-standard` tests. I pushed it into the `misc` package in the meantime, but this didn't seem right, too.

pveentjer replied:
We need to get it fixed. Letting the test run in a warmup phase is flawed by design.

Donnerbart replied:
Danny may need exactly this functionality. We maybe should just rename the class and move it to a distinct package, so it is not misinterpreted as normal test. But I have no idea for a good naming schema for those kind of classes.

Do we have more "tests" like this?

Danny-Hazelcast replied:
Yes I think that is the best idea.  

we have, 3 types of "test"

1  Validation Test  the normal type of test, where we are trying to, catch a memory leak, or stress our system,  through some actions.   in this type of test we don't really care about the throughput or latency of some odd set of operations,  just if it did not crash and we get the result we wanted.

2, performance test,  where all we care is throughput and latency of some set of operations.

3,  system/setup "test",  like this one where the "test" is only really used to create a cluster in some state,  and the "test" part of it  is carried out  in a different way,  say by blocking the ip tables of a member or terminating the box a member is running on,  or creating a higher network load with iperf tool running against one box,  ect ect ect.

Donnerbart commented:
I just tested the most recent version against the old one (both with 2 GB memory, 0.8 usage factor and 10 threads):

`INFO  12:30:43 MapHeapHogTest Waiting for stop completion: 00d 00h 11m 48s`
`INFO  12:14:47 MapHeapHogWarmupTest Waiting for localWarmup completion: 00d 00h 02m 03s`

So quite a nice speedup.

Donnerbart commented:
I removed the "waiting for the cluster" part. This should not be done by any test. If we have problems with the cluster not being up and running, this is a `Coordinator` or `Agent` issue.

The main change of this PR is the 6x speedup in execution (see previous comment).

I moved the test to the package `special` to make clear that it is not a normal test. I totally go with Danny that there is a use case for this non-standard test. So can we merge this?

Donnerbart commented:
\*poke\*
